x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/use_persistent_threat_match_state.ts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useEffect, useRef } from 'react';
+import usePrevious from 'react-use/lib/usePrevious';
+import { isThreatMatchRule } from '../../../../../common/detection_engine/utils';
+import type { FormHook } from '../../../../shared_imports';
+import { useFormData } from '../../../../shared_imports';
+
+interface UsePersistentThreatMatchStateParams {
+  form: FormHook;
+}
+
+interface LastThreatMatchState {
+  threatIndexPatterns: unknown;
+  threatQueryBar: unknown;
+  threatMapping: unknown;
```

Comment:
I don't mind adding concrete types here. My concern is that concrete typing here doesn't bring any value. Instead we get maintenance burden. `useFormData`'s result depends on runtime implementation in particular `UseField` usage and what values were set via `field.setValue`. Due to weak typing in Form lib form values could have different type or being missing completely. In that case it looks like self-deceival having strict types for `useFormData` and `LastThreatMatchState` while it doesn't provide real control on that.

Reply:
I don't think we should neglect types of form data. If you have concern regarding one of the components setting incorrect data, it should be addressed in component. 
Otherwise, we might abandon types for form at all. And that's when we going to have maintenance burden, basically using javascript without any types.

Form data is well typed and we should use it for our advantage: https://github.com/elastic/kibana/blob/8.x/x-pack/solutions/security/plugins/security_solution/public/detections/pages/detection_engine/rules/types.ts#L151-L177

Reply:
I don't think we should neglect types of form data. If you have concern regarding one of the components setting incorrect data, it should be addressed in component. 
Otherwise, we might abandon types for form at all. And that's when we going to have maintenance burden, basically using javascript without any types.

Form data is well typed and we should use it for our advantage: https://github.com/elastic/kibana/blob/8.x/x-pack/solutions/security/plugins/security_solution/public/detections/pages/detection_engine/rules/types.ts#L151-L177